Add one-shot NAS API and  mnasnet based search space.

wanghaoshuang opened the pull request:
Add one-shot NAS.

API doc preview:
https://wanghaoshuang.github.io/PaddleSlim/api/one_shot_api/

Demo doc preview:
https://wanghaoshuang.github.io/PaddleSlim/tutorials/one_shot_nas_demo/

wanghaoshuang committed:
Add one-shot NAS.

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleSlim into one_shot

wanghaoshuang committed:
Fix init token.

qingqing01 commented on demo/one_shot/train.py:
```diff
@@ -0,0 +1,278 @@
+# Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
2019 or 2020  :)

qingqing01 commented on paddleslim/nas/sa_nas.py:
```diff
@@ -173,3 +174,21 @@ def reward(self, score):
                                               self._iter)
+
+
+class OneShotNAS(SANAS):
```
OneShotNAS -> OneShotSANAS ?

ceci3 commented on paddleslim/nas/__init__.py:
```diff
@@ -12,7 +12,11 @@
-__all__ = ['SANAS']
+__all__ = []
+__all__ += sa_nas.__all__
+__all__ += search_space.__all__
```
这样导入包的话 在python3环境下会报错

ceci3 commented on paddleslim/nas/sa_nas.py:
```diff
@@ -137,7 +137,7 @@ def __init__(self,
         return socket.gethostbyname(socket.gethostname())
 
-    def tokens2arch(self, tokens):
+    def tokens2arch(self, tokens=None):
```
这里是不是不给默认值比较好，不然用户也不知道拿到的是什么模型结构

ceci3 commented on paddleslim/nas/search_space/__init__.py:
```diff
@@ -21,10 +21,12 @@
 from .search_space_base import SearchSpaceBase
-
+import one_shot_nas
+from .one_shot_nas import *
```
同上 这样导入包的话 在python3环境下会报错

wanghaoshuang committed:
Refine one-shot API.

wanghaoshuang committed:
Add comments for base class and function in one_shot_nas.

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleSlim into one_shot

wanghaoshuang committed:
Add API docs for one-shot NAS.

wanghaoshuang committed:
Add one-shot demo docs.

wanghaoshuang committed:
Fix copyright.

wanghaoshuang committed:
Fix format.

wanghaoshuang committed:
Refine one shot code.
1. Hide `forward_impl` and `random_tokens`.
2. Fix test_mnist function in demo.
3. Add strategy argument for OneShotSearch function.
4. Remove inference in demo.

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleSlim into one_shot

wanghaoshuang committed:
Fix import of nas module.

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleSlim into one_shot

wanghaoshuang committed:
Fix import of one shot.

wanghaoshuang committed:
Fix api dos.

wanghaoshuang committed:
Support for setting searching steps.

ceci3 reviewed (approved):
LG
